AX: Implement support for NSAccessibilityStringForRangeParameterizedAttribute off the main-thread

twilco opened the pull request:
#### 9ce888f87b027c593caad79000972faa1ac595cf
<pre>
AX: Implement support for NSAccessibilityStringForRangeParameterizedAttribute off the main-thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=293917">https://bugs.webkit.org/show_bug.cgi?id=293917</a>
<a href="https://rdar.apple.com/152449191">rdar://152449191</a>

Reviewed by Joshua Hoffman.

If we convert the NSRange to an AXTextMarkerRange, we can implement NSAccessibilityStringForRangeParameterizedAttribute
off the main-thread -- this commit does so. A few tests needed to be modified to wait for accessibility tree updates
after corresponding main-thread page changes.

* LayoutTests/accessibility/insert-newline-expected.txt:
* LayoutTests/accessibility/insert-newline.html:
* LayoutTests/accessibility/set-selected-text-range-after-newline-expected.txt:
* LayoutTests/accessibility/set-selected-text-range-after-newline.html:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::doAXStringForRange const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(markerRangeFrom):
(computeTextBoundsForRange):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: <a href="https://commits.webkit.org/295725@main">https://commits.webkit.org/295725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/06a5b0e4528a23ceb7f4473e0bffcbdd31bf5f36

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105949 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/25700 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/16094 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/111147 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/56547 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107990 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/26337 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/34203 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/80480 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108955 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/20760 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/95614 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60800 "Passed tests") | 
| | [💥 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/20381 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13712 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55985 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/90144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13748 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113999 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/33089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/24413 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/89559 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/33453 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91843 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/89238 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22747 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/34102 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11917 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/28627 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/33014 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/38425 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32760 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/36109 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/34358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3a9df1436d9624f861ac0c88510c8c4ea36b6592. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46219

twilco commented on Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
```diff
@@ -2866,16 +2866,25 @@ static bool isMatchingPlugin(AXCoreObject& axObject, const AccessibilitySearchCr
+static std::optional<AXTextMarkerRange> markerRangeFrom(NSRange range, const AXCoreObject& object)
+{
+    auto markerToLocation = AXTextMarker { object, 0 }.nextMarkerFromOffset(range.location, ForceSingleOffsetMovement::Yes);
+    auto markerToRangeEnd = markerToLocation.nextMarkerFromOffset(range.length, ForceSingleOffsetMovement::Yes);
```
Great question! I think it is possible for `markerToLocation` to be invalid, since technically ATs can give us any arbitrary NSRange that they want (think `NSRange: {location: 999999999, length: 999999999}`).

My thought process here was that the `if (!markerToRangeEnd.isValid())` checks both markers for validity, as if `markerToLocation` is invalid, creating `markerToRangeEnd` from it will early exit with an invalid marker too. So it's a bit more succinct. But the fact that it caused confusion when you read it (it actually confused me when re-reading this PR a few days ago, and it's my code 😅) means it's probably better to just be clear and have two separate validity checks.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/06a5b0e4528a23ceb7f4473e0bffcbdd31bf5f36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46219

twilco committed:
AX: Implement support for NSAccessibilityStringForRangeParameterizedAttribute off the main-thread
https://bugs.webkit.org/show_bug.cgi?id=293917
rdar://152449191

Reviewed by Joshua Hoffman.

If we convert the NSRange to an AXTextMarkerRange, we can implement NSAccessibilityStringForRangeParameterizedAttribute
off the main-thread -- this commit does so. A few tests needed to be modified to wait for accessibility tree updates
after corresponding main-thread page changes.

* LayoutTests/accessibility/insert-newline-expected.txt:
* LayoutTests/accessibility/insert-newline.html:
* LayoutTests/accessibility/set-selected-text-range-after-newline-expected.txt:
* LayoutTests/accessibility/set-selected-text-range-after-newline.html:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedObject.cpp:
(WebCore::AXIsolatedObject::doAXStringForRange const):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(markerRangeFrom):
(computeTextBoundsForRange):
(-[WebAccessibilityObjectWrapper accessibilityAttributeValue:forParameter:]):

Canonical link: https://commits.webkit.org/295725@main
